### ISSUE TITLE
Empty mesh is allowed under certain conditions

### ISSUE THREAD

**Issue by BenjaminRodenberg** [precice/precice#1158]
**Describe your setup**

I'm currently working on #1029. Here, I wanted to add [this assertion](https://github.com/BenjaminRodenberg/precice/blob/90415267f8a9be0daecf38edd3235801a57e17cd/src/time/Waveform.cpp#L19-L27) which ensures that no empty data is added. Here, I observed that some tests are failing:

1. Serial tests where the first participant provides an empty mesh. (I think this is a problem)
    * `mpirun -np 4 ./testprecice -t PreciceTests/Serial/Time/Explicit/SerialCoupling/DoNothingWithSubcycling`
    * `mpirun -np 4 ./testprecice -t PreciceTests/Serial/testImplicit`
2. Parallel test where one or more ranks do not provide a mesh. (I don't think that this is a problem, because this situation might happen, if a parallel participant has ranks that are creating a `SolverInterface`, but do not contribute to the coupling interface. This is a valid use-case of preCICE, e.g. for the FEniCS adapter). A list of the corresponding tests:
    * `PreciceTests/Parallel/EnforceGatherScatterEmptyMaster`
    * `PreciceTests/Parellel/TestQN*EmptyPartition`
    * `PreciceTests/Parellel/AccessReceivedMeshEmptyPartition`
    * `PreciceTests/Parellel/AccessReceivedMeshEmptyPartitionTwoLevelInit`
    * `PreciceTests/Parellel/NearestProjectionRePartitioning`

**Describe the problem**

I will now concentrate on class (1) mentioned above: To me this looks like a problem, because the empty mesh is allowed, if the first participant has an empty mesh, but If I remove the mesh for the second participant in `DoNothingWithSubcycling` I get the following error:

```
12:09:28.232530|SolverTwo|0|partition::ProvidedPartition|l109|communicate|ERROR: The provided mesh "Test-Square" is empty. Please set the mesh using setMeshXXX() prior to calling initialize().
```

**Step To Reproduce**

Uncomment the assertion and run the tests under revision 90415267f8a9be0daecf38edd3235801a57e17cd.

**Expected behaviour**

I expect preCICE to raise an error, if a participant provides an empty mesh (first and second). This is mainly what I would like to discuss in this issue.

**Additional context**

* Some of the parallel tests mentioned above were added here: https://github.com/precice/precice/pull/905 (took me some time to figure this out, because the history is a bit messed up, so I will provide it here)
* I decided to fix the serial tests by providing mesh vertices for both participants. The parallel tests are untouched.
* In #1029 I decided to remove the assertion, because it is too strict.

**Comment by uekerman** [comment]
Good point. Yes, I think currently, we only raise an error when an empty provided mesh is communicated. If it is not communicated, then no error is thrown. We should add a check also for this case. Empty here means that all ranks are empty.

**Comment by BenjaminRodenberg** [comment]
Another situation where this bug is visible:

`SolverOne.py`:
```
import precice
solverInterface = precice.Interface("SolverOne", "precice-config.xml", 0,1)
#solverInterface.initialize()
# waiting for SolverTwo's solverInterface.initialize() 
# ERROR: The provided mesh "MeshOne" is empty. Please set the mesh using setMeshXXX() prior to calling initialize().
# -> Ok, let's provide a mesh
meshOneId = solverInterface.get_mesh_id("MeshOne")
vertexId = solverInterface.set_mesh_vertex(meshOneId, [1,2,3])
solverInterface.initialize()
# waiting for SolverTwo's solverInterface.initialize()
# all good.
```

`SolverTwo.py`:
```
import precice
solverInterface = precice.Interface("SolverTwo", "precice-config.xml", 0,1)
solverInterface.initialize()
# waiting for SolverOne's solverInterface.initialize() 
# no error, but would expect one, because MeshTwo is not provided by SolverTwo
# expected: ERROR: The provided mesh "MeshTwo" is empty. Please set the mesh using setMeshXXX() prior to calling initialize().
```

`precice-config.xml`:
```
<?xml version="1.0"?>

<precice-configuration>
  <solver-interface dimensions="3" >
   
    <data:vector name="dataOne"  />
    <data:vector name="dataTwo"  />

    <mesh name="MeshOne">
      <use-data name="dataOne" />
      <use-data name="dataTwo" />
    </mesh>

    <mesh name="MeshTwo">
      <use-data name="dataOne" />
      <use-data name="dataTwo" />
    </mesh>

    <participant name="SolverOne">
      <use-mesh name="MeshOne" provide="yes"/>
      <write-data name="dataOne" mesh="MeshOne" />
      <read-data  name="dataTwo" mesh="MeshOne" />
    </participant>

    <participant name="SolverTwo">
      <use-mesh name="MeshOne" from="SolverOne"/>
      <use-mesh name="MeshTwo" provide="yes"/>
      <mapping:nearest-neighbor   direction="write" from="MeshTwo" to="MeshOne" constraint="conservative"/>
      <mapping:nearest-neighbor direction="read"  from="MeshOne" to="MeshTwo" constraint="consistent" />
      <write-data name="dataTwo" mesh="MeshTwo" />
      <read-data  name="dataOne" mesh="MeshTwo" />
    </participant>

    <m2n:sockets from="SolverOne" to="SolverTwo"/>

    <coupling-scheme:parallel-implicit> 
      <participants first="SolverOne" second="SolverTwo" /> 
      <max-time-windows value="2" />
      <time-window-size value="1.0" />
      <max-iterations value="2" />
      <min-iteration-convergence-measure min-iterations="5" data="dataOne" mesh="MeshOne"/>
      <exchange data="dataOne" mesh="MeshOne" from="SolverOne" to="SolverTwo" />
      <exchange data="dataTwo" mesh="MeshOne" from="SolverTwo" to="SolverOne"/>
    </coupling-scheme:parallel-implicit>                           
  </solver-interface>

</precice-configuration>
```

**Comment by BenjaminRodenberg** [comment]
Shifting it to `3.x.x`, because it is not critical in my opinion. Feel free to shift it back to `3.0.0`, if you are willing to work on it.

**Comment by Tanisha127** [comment]
Hi @fsimonis and @BenjaminRodenberg 

I would like to work on this issue.
From the discussion, I understand that we should add a validation check to raise an error when all ranks of a provided mesh are empty, even if the mesh is not communicated.
Could you please confirm whether this is still the intended behavior before I start implementing it?

Thanks!
